# Building a Relational Database Part 2

- toc: true
- badges: true
- comments: true
- image: images/PostgresSQL.png

## Using DDL, create each of the relations in the postgres server. 

In [5]:
import getpass
mypasswd = getpass.getpass()
username = 'dlr6w3'
host = 'pgsql.dsa.lan'
database = 'dsa_student'

········


In [6]:
##Create connection
from sqlalchemy.engine.url import URL
from sqlalchemy import create_engine

postgres_db = {'drivername': 'postgres',
               'username': username,
               'password': mypasswd,
               'host': host,
               'database' :database}
engine = create_engine(URL(**postgres_db), echo=True)
del mypasswd

In [7]:
query = """

DROP TABLE IF EXISTS crime CASCADE;

CREATE TABLE crime(
    crime_id     INT NOT NULL UNIQUE,
    iucr         VARCHAR(10) NOT NULL,
    fbi          VARCHAR(10) NOT NULL,
    primary_desc VARCHAR(50) NOT NULL,
    description  VARCHAR(100) NOT NULL,
    PRIMARY KEY (crime_id)
) 


"""

with engine.connect() as connection:
    res = connection.execute(query)
    print(res)

2020-12-18 09:56:14,847 INFO sqlalchemy.engine.base.Engine select version()
2020-12-18 09:56:14,848 INFO sqlalchemy.engine.base.Engine {}
2020-12-18 09:56:14,850 INFO sqlalchemy.engine.base.Engine select current_schema()
2020-12-18 09:56:14,851 INFO sqlalchemy.engine.base.Engine {}
2020-12-18 09:56:14,853 INFO sqlalchemy.engine.base.Engine SELECT CAST('test plain returns' AS VARCHAR(60)) AS anon_1
2020-12-18 09:56:14,854 INFO sqlalchemy.engine.base.Engine {}
2020-12-18 09:56:14,855 INFO sqlalchemy.engine.base.Engine SELECT CAST('test unicode returns' AS VARCHAR(60)) AS anon_1
2020-12-18 09:56:14,855 INFO sqlalchemy.engine.base.Engine {}
2020-12-18 09:56:14,856 INFO sqlalchemy.engine.base.Engine show standard_conforming_strings
2020-12-18 09:56:14,857 INFO sqlalchemy.engine.base.Engine {}
2020-12-18 09:56:14,858 INFO sqlalchemy.engine.base.Engine 

DROP TABLE IF EXISTS crime CASCADE;

CREATE TABLE crime(
    crime_id     INT NOT NULL UNIQUE,
    iucr         VARCHAR(10) NOT NULL,
    fb

In [8]:
query = """

DROP TABLE IF EXISTS location CASCADE;

CREATE TABLE location (
    location_id             INT NOT NULL UNIQUE,
    location_description    VARCHAR(50) NOT NULL,
    block                   VARCHAR(50) NOT NULL,
    beat                    INT NOT NULL,
    district                INT NOT NULL,
    ward                    FLOAT NOT NULL,
    community_area          FLOAT NOT NULL,
    latitude                FLOAT NOT NULL,
    longitude               FLOAT NOT NULL,
    x_coordinate            FLOAT NOT NULL,
    y_coordinate            FLOAT NOT NULL,
    PRIMARY KEY (location_id)
    
)

"""

with engine.connect() as connection:
    res = connection.execute(query)
    print(res)


2020-12-18 09:56:22,560 INFO sqlalchemy.engine.base.Engine 

DROP TABLE IF EXISTS location CASCADE;

CREATE TABLE location (
    location_id             INT NOT NULL UNIQUE,
    location_description    VARCHAR(50) NOT NULL,
    block                   VARCHAR(50) NOT NULL,
    beat                    INT NOT NULL,
    district                INT NOT NULL,
    ward                    FLOAT NOT NULL,
    community_area          FLOAT NOT NULL,
    latitude                FLOAT NOT NULL,
    longitude               FLOAT NOT NULL,
    x_coordinate            FLOAT NOT NULL,
    y_coordinate            FLOAT NOT NULL,
    PRIMARY KEY (location_id)
    
)


2020-12-18 09:56:22,561 INFO sqlalchemy.engine.base.Engine {}
2020-12-18 09:56:22,574 INFO sqlalchemy.engine.base.Engine COMMIT


In [10]:
query = """

DROP TABLE IF EXISTS crime_report CASCADE;

CREATE TABLE crime_report(
    case_id        VARCHAR(12) NOT NULL UNIQUE,
    case_number    VARCHAR(12) NOT NULL,
    crime_id       INT NOT NULL,
    location_id    INT NOT NULL,
    arrest         BOOL NOT NULL,
    domestic       BOOL NOT NULL,
    case_date      TIMESTAMP NOT NULL,
    case_year      INT NOT NULL,
    updated_on     TIMESTAMP NOT NULL,
    PRIMARY KEY (case_id),
    FOREIGN KEY (crime_id) REFERENCES crime(crime_id),
    FOREIGN KEY (location_id) REFERENCES location(location_id)


)

"""

with engine.connect() as connection:
    res = connection.execute(query)
    print(res)

2020-12-18 09:57:47,375 INFO sqlalchemy.engine.base.Engine 

DROP TABLE IF EXISTS crime_report CASCADE;

CREATE TABLE crime_report(
    case_id        VARCHAR(12) NOT NULL UNIQUE,
    case_number    VARCHAR(12) NOT NULL,
    crime_id       INT NOT NULL,
    location_id    INT NOT NULL,
    arrest         BOOL NOT NULL,
    domestic       BOOL NOT NULL,
    case_date      TIMESTAMP NOT NULL,
    case_year      INT NOT NULL,
    updated_on     TIMESTAMP NOT NULL,
    PRIMARY KEY (case_id),
    FOREIGN KEY (crime_id) REFERENCES crime(crime_id),
    FOREIGN KEY (location_id) REFERENCES location(location_id)


)


2020-12-18 09:57:47,376 INFO sqlalchemy.engine.base.Engine {}
2020-12-18 09:57:47,388 INFO sqlalchemy.engine.base.Engine COMMIT


## Show the table definitions using psql or querying information_schema.colums catalog


Crime Table:

dsa_student=> \d+ crime
                                              Table "dlr6w3.crime"
    Column    |          Type          | Collation | Nullable | Default | Storage  | Stats target | Description
--------------+------------------------+-----------+----------+---------+----------+--------------+-------------
 crime_id     | integer                |           | not null |         | plain    |              |
 iucr         | character varying(10)  |           | not null |         | extended |              |
 fbi          | character varying(10)  |           | not null |         | extended |              |
 primary_desc | character varying(50)  |           | not null |         | extended |              |
 description  | character varying(100) |           | not null |         | extended |              |
Indexes:
    "crime_pkey" PRIMARY KEY, btree (crime_id)
Referenced by:
    TABLE "crime_report" CONSTRAINT "crime_report_crime_id_fkey" FOREIGN KEY (crime_id) REFERENCES crime(crime_id)

Location Table:

dsa_student=> \d+ location
                                                Table "dlr6w3.location"
        Column        |         Type          | Collation | Nullable | Default | Storage  | Stats target | Description
----------------------+-----------------------+-----------+----------+---------+----------+--------------+-------------
 location_id          | integer               |           | not null |         | plain    |              |
 location_description | character varying(50) |           | not null |         | extended |              |
 block                | character varying(50) |           | not null |         | extended |              |
 beat                 | integer               |           | not null |         | plain    |              |
 district             | integer               |           | not null |         | plain    |              |
 ward                 | double precision      |           | not null |         | plain    |              |
 community_area       | double precision      |           | not null |         | plain    |              |
 latitude             | double precision      |           | not null |         | plain    |              |
 longitude            | double precision      |           | not null |         | plain    |              |
 x_coordinate         | double precision      |           | not null |         | plain    |              |
 y_coordinate         | double precision      |           | not null |         | plain    |              |
Indexes:
    "location_pkey" PRIMARY KEY, btree (location_id)
Referenced by:
    TABLE "crime_report" CONSTRAINT "crime_report_location_id_fkey" FOREIGN KEY (location_id) REFERENCES location(location_id)
 
Crime_Report Table: 
 
dsa_student=> \d+ crime_report
                                            Table "dlr6w3.crime_report"
   Column    |            Type             | Collation | Nullable | Default | Storage  | Stats target | Description
-------------+-----------------------------+-----------+----------+---------+----------+--------------+-------------
 case_id     | character varying(12)       |           | not null |         | extended |              |
 case_number | character varying(12)       |           | not null |         | extended |              |
 crime_id    | integer                     |           | not null |         | plain    |              |
 location_id | integer                     |           | not null |         | plain    |              |
 arrest      | boolean                     |           | not null |         | plain    |              |
 domestic    | boolean                     |           | not null |         | plain    |              |
 case_date   | timestamp without time zone |           | not null |         | plain    |              |
 case_year   | integer                     |           | not null |         | plain    |              |
 updated_on  | timestamp without time zone |           | not null |         | plain    |              |
Indexes:
    "crime_report_pkey" PRIMARY KEY, btree (case_id)
Foreign-key constraints:
    "crime_report_crime_id_fkey" FOREIGN KEY (crime_id) REFERENCES crime(crime_id)
    "crime_report_location_id_fkey" FOREIGN KEY (location_id) REFERENCES location(location_id)

## Load the data from the given csv file to the relations


In [11]:
##Load the csv
import pandas as pd
datapath = "/dsa/data/DSA-7030/Chicago-Crime-Sample-2012-updated.csv"
df = pd.read_csv(datapath, index_col=0)

In [12]:
df.head()

Case Number                    Date               Block  IUCR  \
ID                                                                       
12236659    JD449284  11/27/2012 06:00:00 PM  028XX W BELDEN AVE  0266   
12214909    JD424223  09/15/2012 01:30:00 PM   045XX W CERMAK RD  0620   
12235566    JD447838  01/01/2012 12:01:00 AM     031XX W 44TH ST  1752   
12234792    JD447109  12/07/2012 12:00:00 PM   107XX S MORGAN ST  1153   
12230182    JD441224  02/27/2012 06:50:00 PM  017XX W BERWYN AVE  1153   

                        Primary Type  \
ID                                     
12236659     CRIMINAL SEXUAL ASSAULT   
12214909                    BURGLARY   
12235566  OFFENSE INVOLVING CHILDREN   
12234792          DECEPTIVE PRACTICE   
12230182          DECEPTIVE PRACTICE   

                                                Description  \
ID                                                            
12236659                                          PREDATORY   
12214909                                     UNLAWFUL ENTRY   
12235566  AGGRAVATED CRIMINAL SEXUAL ABUSE BY FAMILY MEMBER   
12234792                FINANCIAL IDENTITY THEFT OVER $ 300   
12230182                FINANCIAL IDENTITY THEFT OVER $ 300   

         Location Description  Arrest  Domestic  Beat  ...  Ward  \
ID                                                     ...         
12236659            RESIDENCE   False     False  1414  ...   1.0   
12214909      OTHER (SPECIFY)   False     False  1012  ...  24.0   
12235566            RESIDENCE   False     False   922  ...  15.0   
12234792            RESIDENCE   False     False  2234  ...  34.0   
12230182                  NaN   False     False  2012  ...  40.0   

          Community Area  FBI Code X Coordinate  Y Coordinate  Year  \
ID                                                                    
12236659            22.0        02          NaN           NaN  2012   
12214909            29.0        05          NaN           NaN  2012   
12235566            58.0        17          NaN           NaN  2012   
12234792            75.0        11          NaN           NaN  2012   
12230182            77.0        11          NaN           NaN  2012   

                      Updated On Latitude  Longitude  Location  
ID                                                              
12236659  12/05/2020 03:46:44 PM      NaN        NaN       NaN  
12214909  12/05/2020 03:44:06 PM      NaN        NaN       NaN  
12235566  12/03/2020 03:51:39 PM      NaN        NaN       NaN  
12234792  12/02/2020 03:56:38 PM      NaN        NaN       NaN  
12230182  11/26/2020 03:48:35 PM      NaN        NaN       NaN  

[5 rows x 21 columns]

In [13]:
## Look at dimensions and Count NaNs
print(df.shape)
df.isna().sum()

(336149, 21)


Case Number               1
Date                      0
Block                     0
IUCR                      0
Primary Type              0
Description               0
Location Description    447
Arrest                    0
Domestic                  0
Beat                      0
District                  0
Ward                      7
Community Area           26
FBI Code                  0
X Coordinate            703
Y Coordinate            703
Year                      0
Updated On                0
Latitude                703
Longitude               703
Location                703
dtype: int64

In [14]:
##Remove nans with values to not violate NOT NULL
df['Case Number'] = df['Case Number'].where(pd.notnull(df['Case Number']), -1)
df['Location Description'] = df['Location Description'].where(pd.notnull(df['Location Description']), 'NA')
df['Ward'] = df['Ward'].where(pd.notnull(df['Ward']), -1.00)
df['Community Area'] = df['Community Area'].where(pd.notnull(df['Community Area']), -1.00)
df['X Coordinate'] = df['X Coordinate'].where(pd.notnull(df['X Coordinate']), -1.00)
df['Y Coordinate'] = df['Y Coordinate'].where(pd.notnull(df['Y Coordinate']), -1.00)
df['Latitude'] = df['Latitude'].where(pd.notnull(df['Latitude']), -1.00)
df['Longitude'] = df['Longitude'].where(pd.notnull(df['Longitude']), -1.00)
df['Location'] = df['Location'].where(pd.notnull(df['Location']), 'NA')


In [15]:
## Look at updated dimensions and check if NaNs are gone
print(df.shape)
df.isna().sum()

(336149, 21)


Case Number             0
Date                    0
Block                   0
IUCR                    0
Primary Type            0
Description             0
Location Description    0
Arrest                  0
Domestic                0
Beat                    0
District                0
Ward                    0
Community Area          0
FBI Code                0
X Coordinate            0
Y Coordinate            0
Year                    0
Updated On              0
Latitude                0
Longitude               0
Location                0
dtype: int64

In [16]:
##Create crime dataframe
crime_df=pd.DataFrame({
                        'IUCR':  df.IUCR,
                        'FBI Code': df['FBI Code'],
                        'Primary Type': df['Primary Type'],
                        'Description':  df.Description,
                    
                    
})
print(crime_df.shape)
crime_df.head()


(336149, 4)


IUCR FBI Code                Primary Type  \
ID                                                    
12236659  0266       02     CRIMINAL SEXUAL ASSAULT   
12214909  0620       05                    BURGLARY   
12235566  1752       17  OFFENSE INVOLVING CHILDREN   
12234792  1153       11          DECEPTIVE PRACTICE   
12230182  1153       11          DECEPTIVE PRACTICE   

                                                Description  
ID                                                           
12236659                                          PREDATORY  
12214909                                     UNLAWFUL ENTRY  
12235566  AGGRAVATED CRIMINAL SEXUAL ABUSE BY FAMILY MEMBER  
12234792                FINANCIAL IDENTITY THEFT OVER $ 300  
12230182                FINANCIAL IDENTITY THEFT OVER $ 300

In [17]:
##Get unique rows
crime_df_unique = crime_df.drop_duplicates()
print(crime_df_unique.shape)
crime_df_unique.head()

(339, 4)


IUCR FBI Code                Primary Type  \
ID                                                    
12236659  0266       02     CRIMINAL SEXUAL ASSAULT   
12214909  0620       05                    BURGLARY   
12235566  1752       17  OFFENSE INVOLVING CHILDREN   
12234792  1153       11          DECEPTIVE PRACTICE   
8732938   0281       02     CRIMINAL SEXUAL ASSAULT   

                                                Description  
ID                                                           
12236659                                          PREDATORY  
12214909                                     UNLAWFUL ENTRY  
12235566  AGGRAVATED CRIMINAL SEXUAL ABUSE BY FAMILY MEMBER  
12234792                FINANCIAL IDENTITY THEFT OVER $ 300  
8732938                                      NON-AGGRAVATED

In [18]:
##Create Id column
crime_df_unique['crime_id'] = range(0, len(crime_df_unique.IUCR))
crime_df_unique.tail()

/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


IUCR FBI Code                Primary Type  \
ID                                                   
8946827  2091       18                   NARCOTICS   
8948914  2895       26      PUBLIC PEACE VIOLATION   
8950610  1531       16                PROSTITUTION   
9420694  1710       20  OFFENSE INVOLVING CHILDREN   
9762752  142B       15           WEAPONS VIOLATION   

                           Description  crime_id  
ID                                                
8946827               FORFEIT PROPERTY       334  
8948914   INTERFERE W/ EMERGENCY EQUIP       335  
8950610               JUVENILE PIMPING       336  
9420694  ENDANGERING LIFE/HEALTH CHILD       337  
9762752    UNLAWFUL SALE OTHER FIREARM       338

In [19]:
##Create location df

location_df=pd.DataFrame({
                        
                'Location Description':   df['Location Description'] ,
                'Block':                  df.Block ,
                'Beat':                    df.Beat,
                'District':               df.District ,
                'Ward':                  df.Ward  ,
                'Community Area':          df['Community Area'],
                'Latitude':               df.Latitude ,
                'Longitude':              df.Longitude ,
                'X Coordinate':            df['X Coordinate'],
                'Y Coordinate':          df['Y Coordinate'] ,  
                    
})
print(location_df.shape)
location_df.head()

(336149, 10)


Location Description               Block  Beat  District  Ward  \
ID                                                                        
12236659            RESIDENCE  028XX W BELDEN AVE  1414        14   1.0   
12214909      OTHER (SPECIFY)   045XX W CERMAK RD  1012        10  24.0   
12235566            RESIDENCE     031XX W 44TH ST   922         9  15.0   
12234792            RESIDENCE   107XX S MORGAN ST  2234        22  34.0   
12230182                   NA  017XX W BERWYN AVE  2012        20  40.0   

          Community Area  Latitude  Longitude  X Coordinate  Y Coordinate  
ID                                                                         
12236659            22.0      -1.0       -1.0          -1.0          -1.0  
12214909            29.0      -1.0       -1.0          -1.0          -1.0  
12235566            58.0      -1.0       -1.0          -1.0          -1.0  
12234792            75.0      -1.0       -1.0          -1.0          -1.0  
12230182            77.0      -1.0       -1.0          -1.0          -1.0

In [20]:
##Get unique rows
location_df_unique = location_df.drop_duplicates()
print(location_df_unique.shape)
location_df_unique.head()

(244517, 10)


Location Description               Block  Beat  District  Ward  \
ID                                                                        
12236659            RESIDENCE  028XX W BELDEN AVE  1414        14   1.0   
12214909      OTHER (SPECIFY)   045XX W CERMAK RD  1012        10  24.0   
12235566            RESIDENCE     031XX W 44TH ST   922         9  15.0   
12234792            RESIDENCE   107XX S MORGAN ST  2234        22  34.0   
12230182                   NA  017XX W BERWYN AVE  2012        20  40.0   

          Community Area  Latitude  Longitude  X Coordinate  Y Coordinate  
ID                                                                         
12236659            22.0      -1.0       -1.0          -1.0          -1.0  
12214909            29.0      -1.0       -1.0          -1.0          -1.0  
12235566            58.0      -1.0       -1.0          -1.0          -1.0  
12234792            75.0      -1.0       -1.0          -1.0          -1.0  
12230182            77.0      -1.0       -1.0          -1.0          -1.0

In [21]:
##Create Id column
location_df_unique['Location_ID'] = range(0, len(location_df_unique.Block))
print(location_df_unique.shape)
location_df_unique.tail()

(244517, 11)


/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


Location Description                   Block  Beat  District  Ward  \
ID                                                                           
9984444            APARTMENT      076XX S COLFAX AVE   421         4   7.0   
9987208               STREET       067XX S PERRY AVE   722         7   6.0   
9990541            RESIDENCE       010XX N HONORE ST  1212        12   1.0   
9991722            APARTMENT       019XX S DRAKE AVE  1024        10  24.0   
9998136            RESIDENCE  045XX W WRIGHTWOOD AVE  2521        25  31.0   

         Community Area   Latitude  Longitude  X Coordinate  Y Coordinate  \
ID                                                                          
9984444            43.0  41.756505 -87.561474     1194850.0     1854806.0   
9987208            69.0  41.772012 -87.628344     1176558.0     1860300.0   
9990541            24.0  41.900421 -87.673514     1163877.0     1906994.0   
9991722            29.0  41.854248 -87.713694     1153060.0     1890087.0   
9998136            20.0  41.928606 -87.740772     1145496.0     1917131.0   

         Location_ID  
ID                    
9984444       244512  
9987208       244513  
9990541       244514  
9991722       244515  
9998136       244516

In [22]:
##Join data to original df to add location and crime ids
join1 =  pd.merge(df, location_df_unique, how='left', on=['Location Description', 'Block', 'Beat', 'District', 'Ward', 'Community Area', 'Community Area', 'Latitude', 'Longitude', 'X Coordinate', 'Y Coordinate'])

In [23]:
print(join1.shape)
join1.tail()

(336149, 22)


Case Number                    Date                   Block  IUCR  \
336144    HY173771  10/11/2012 06:00:00 AM      076XX S COLFAX AVE  1153   
336145    HY177204  12/15/2012 09:00:00 AM       067XX S PERRY AVE  0910   
336146    HY180702  02/07/2012 09:00:00 AM       010XX N HONORE ST  1153   
336147    HY181590  04/01/2012 12:01:00 AM       019XX S DRAKE AVE  1153   
336148    HY188115  01/01/2012 09:00:00 AM  045XX W WRIGHTWOOD AVE  1153   

               Primary Type                          Description  \
336144   DECEPTIVE PRACTICE  FINANCIAL IDENTITY THEFT OVER $ 300   
336145  MOTOR VEHICLE THEFT                           AUTOMOBILE   
336146   DECEPTIVE PRACTICE  FINANCIAL IDENTITY THEFT OVER $ 300   
336147   DECEPTIVE PRACTICE  FINANCIAL IDENTITY THEFT OVER $ 300   
336148   DECEPTIVE PRACTICE  FINANCIAL IDENTITY THEFT OVER $ 300   

       Location Description  Arrest  Domestic  Beat  ...  Community Area  \
336144            APARTMENT   False     False   421  ...            43.0   
336145               STREET   False     False   722  ...            69.0   
336146            RESIDENCE   False     False  1212  ...            24.0   
336147            APARTMENT   False     False  1024  ...            29.0   
336148            RESIDENCE   False     False  2521  ...            20.0   

        FBI Code  X Coordinate Y Coordinate  Year              Updated On  \
336144        11     1194850.0    1854806.0  2012  02/10/2018 03:50:01 PM   
336145        07     1176558.0    1860300.0  2012  02/10/2018 03:50:01 PM   
336146        11     1163877.0    1906994.0  2012  02/09/2018 03:44:29 PM   
336147        11     1153060.0    1890087.0  2012  02/09/2018 03:44:29 PM   
336148        11     1145496.0    1917131.0  2012  02/09/2018 03:44:29 PM   

         Latitude  Longitude                       Location  Location_ID  
336144  41.756505 -87.561474  (41.756504746, -87.561473501)       244512  
336145  41.772012 -87.628344  (41.772011702, -87.628344479)       244513  
336146  41.900421 -87.673514  (41.900420776, -87.673514317)       244514  
336147  41.854248 -87.713694  (41.854247595, -87.713694133)       244515  
336148  41.928606 -87.740772  (41.928605829, -87.740771814)       244516  

[5 rows x 22 columns]

In [24]:
join2 = pd.merge(join1, crime_df_unique, how='left', on=['IUCR', 'FBI Code', 'Primary Type', 'Description'])


In [25]:
print(join2.shape)
join2.head()

(336149, 23)


Case Number                    Date               Block  IUCR  \
0    JD449284  11/27/2012 06:00:00 PM  028XX W BELDEN AVE  0266   
1    JD424223  09/15/2012 01:30:00 PM   045XX W CERMAK RD  0620   
2    JD447838  01/01/2012 12:01:00 AM     031XX W 44TH ST  1752   
3    JD447109  12/07/2012 12:00:00 PM   107XX S MORGAN ST  1153   
4    JD441224  02/27/2012 06:50:00 PM  017XX W BERWYN AVE  1153   

                 Primary Type  \
0     CRIMINAL SEXUAL ASSAULT   
1                    BURGLARY   
2  OFFENSE INVOLVING CHILDREN   
3          DECEPTIVE PRACTICE   
4          DECEPTIVE PRACTICE   

                                         Description Location Description  \
0                                          PREDATORY            RESIDENCE   
1                                     UNLAWFUL ENTRY      OTHER (SPECIFY)   
2  AGGRAVATED CRIMINAL SEXUAL ABUSE BY FAMILY MEMBER            RESIDENCE   
3                FINANCIAL IDENTITY THEFT OVER $ 300            RESIDENCE   
4                FINANCIAL IDENTITY THEFT OVER $ 300                   NA   

   Arrest  Domestic  Beat  ...  FBI Code  X Coordinate  Y Coordinate  Year  \
0   False     False  1414  ...        02          -1.0          -1.0  2012   
1   False     False  1012  ...        05          -1.0          -1.0  2012   
2   False     False   922  ...        17          -1.0          -1.0  2012   
3   False     False  2234  ...        11          -1.0          -1.0  2012   
4   False     False  2012  ...        11          -1.0          -1.0  2012   

               Updated On  Latitude  Longitude Location  Location_ID  crime_id  
0  12/05/2020 03:46:44 PM      -1.0       -1.0       NA            0         0  
1  12/05/2020 03:44:06 PM      -1.0       -1.0       NA            1         1  
2  12/03/2020 03:51:39 PM      -1.0       -1.0       NA            2         2  
3  12/02/2020 03:56:38 PM      -1.0       -1.0       NA            3         3  
4  11/26/2020 03:48:35 PM      -1.0       -1.0       NA            4         3  

[5 rows x 23 columns]

In [26]:
##Add case id
join2['case_id'] = range(0, len(join2['Case Number']))
join2.tail()

Case Number                    Date                   Block  IUCR  \
336144    HY173771  10/11/2012 06:00:00 AM      076XX S COLFAX AVE  1153   
336145    HY177204  12/15/2012 09:00:00 AM       067XX S PERRY AVE  0910   
336146    HY180702  02/07/2012 09:00:00 AM       010XX N HONORE ST  1153   
336147    HY181590  04/01/2012 12:01:00 AM       019XX S DRAKE AVE  1153   
336148    HY188115  01/01/2012 09:00:00 AM  045XX W WRIGHTWOOD AVE  1153   

               Primary Type                          Description  \
336144   DECEPTIVE PRACTICE  FINANCIAL IDENTITY THEFT OVER $ 300   
336145  MOTOR VEHICLE THEFT                           AUTOMOBILE   
336146   DECEPTIVE PRACTICE  FINANCIAL IDENTITY THEFT OVER $ 300   
336147   DECEPTIVE PRACTICE  FINANCIAL IDENTITY THEFT OVER $ 300   
336148   DECEPTIVE PRACTICE  FINANCIAL IDENTITY THEFT OVER $ 300   

       Location Description  Arrest  Domestic  Beat  ...  X Coordinate  \
336144            APARTMENT   False     False   421  ...     1194850.0   
336145               STREET   False     False   722  ...     1176558.0   
336146            RESIDENCE   False     False  1212  ...     1163877.0   
336147            APARTMENT   False     False  1024  ...     1153060.0   
336148            RESIDENCE   False     False  2521  ...     1145496.0   

        Y Coordinate  Year              Updated On   Latitude  Longitude  \
336144     1854806.0  2012  02/10/2018 03:50:01 PM  41.756505 -87.561474   
336145     1860300.0  2012  02/10/2018 03:50:01 PM  41.772012 -87.628344   
336146     1906994.0  2012  02/09/2018 03:44:29 PM  41.900421 -87.673514   
336147     1890087.0  2012  02/09/2018 03:44:29 PM  41.854248 -87.713694   
336148     1917131.0  2012  02/09/2018 03:44:29 PM  41.928606 -87.740772   

                             Location Location_ID  crime_id  case_id  
336144  (41.756504746, -87.561473501)      244512         3   336144  
336145  (41.772011702, -87.628344479)      244513        26   336145  
336146  (41.900420776, -87.673514317)      244514         3   336146  
336147  (41.854247595, -87.713694133)      244515         3   336147  
336148  (41.928605829, -87.740771814)      244516         3   336148  

[5 rows x 24 columns]

In [27]:
##Create crime_report df

creport_df_sql=pd.DataFrame({
                'case_id': join2.case_id,
                'case_number':  join2['Case Number']  ,
                'crime_id': join2['crime_id'],
                'location_id': join2['Location_ID'],
                'arrest':       join2.Arrest  ,
                'domestic':     join2.Domestic  ,
                'case_date':    join2.Date  ,
                'case_year':     join2.Year ,
                'updated_on':    join2['Updated On'] ,       
                    
                    
})
print(creport_df_sql.shape)
creport_df_sql.head()



(336149, 9)


case_id case_number  crime_id  location_id  arrest  domestic  \
0        0    JD449284         0            0   False     False   
1        1    JD424223         1            1   False     False   
2        2    JD447838         2            2   False     False   
3        3    JD447109         3            3   False     False   
4        4    JD441224         3            4   False     False   

                case_date  case_year              updated_on  
0  11/27/2012 06:00:00 PM       2012  12/05/2020 03:46:44 PM  
1  09/15/2012 01:30:00 PM       2012  12/05/2020 03:44:06 PM  
2  01/01/2012 12:01:00 AM       2012  12/03/2020 03:51:39 PM  
3  12/07/2012 12:00:00 PM       2012  12/02/2020 03:56:38 PM  
4  02/27/2012 06:50:00 PM       2012  11/26/2020 03:48:35 PM

In [28]:
##change the two date columns to timestamp
creport_df_sql.case_date = pd.to_datetime(creport_df_sql.case_date, infer_datetime_format=True) 
creport_df_sql.updated_on = pd.to_datetime(creport_df_sql.updated_on, infer_datetime_format=True)

In [29]:
##check NAs
print(creport_df_sql.isna().sum())
print(location_df_unique.isna().sum())
print(crime_df_unique.isna().sum())

case_id        0
case_number    0
crime_id       0
location_id    0
arrest         0
domestic       0
case_date      0
case_year      0
updated_on     0
dtype: int64
Location Description    0
Block                   0
Beat                    0
District                0
Ward                    0
Community Area          0
Latitude                0
Longitude               0
X Coordinate            0
Y Coordinate            0
Location_ID             0
dtype: int64
IUCR            0
FBI Code        0
Primary Type    0
Description     0
crime_id        0
dtype: int64


In [30]:
##create sql crime df
crime_df_sql = pd.DataFrame({
                        'crime_id':  crime_df_unique['crime_id'],
                        'iucr':  crime_df_unique.IUCR,
                        'fbi': crime_df_unique['FBI Code'],
                        'primary_desc': crime_df_unique['Primary Type'],
                        'description':  crime_df_unique.Description,
                    
                    
})
print(crime_df_sql.shape)
crime_df_sql.head()

(339, 5)


crime_id  iucr fbi                primary_desc  \
ID                                                         
12236659         0  0266  02     CRIMINAL SEXUAL ASSAULT   
12214909         1  0620  05                    BURGLARY   
12235566         2  1752  17  OFFENSE INVOLVING CHILDREN   
12234792         3  1153  11          DECEPTIVE PRACTICE   
8732938          4  0281  02     CRIMINAL SEXUAL ASSAULT   

                                                description  
ID                                                           
12236659                                          PREDATORY  
12214909                                     UNLAWFUL ENTRY  
12235566  AGGRAVATED CRIMINAL SEXUAL ABUSE BY FAMILY MEMBER  
12234792                FINANCIAL IDENTITY THEFT OVER $ 300  
8732938                                      NON-AGGRAVATED

In [31]:
##create sql location df
location_df_sql = pd.DataFrame({
                        
                'location_id':            location_df_unique['Location_ID'],
                'location_description':   location_df_unique['Location Description'] ,
                'block':                  location_df_unique.Block ,
                'beat':                    location_df_unique.Beat,
                'district':               location_df_unique.District ,
                'ward':                  location_df_unique.Ward  ,
                'community_area':          location_df_unique['Community Area'],
                'latitude':               location_df_unique.Latitude ,
                'longitude':              location_df_unique.Longitude ,
                'x_coordinate':            location_df_unique['X Coordinate'],
                'y_coordinate':          location_df_unique['Y Coordinate'] ,  
                    
})
print(location_df_sql.shape)
location_df_sql.head()

(244517, 11)


location_id location_description               block  beat  \
ID                                                                     
12236659            0            RESIDENCE  028XX W BELDEN AVE  1414   
12214909            1      OTHER (SPECIFY)   045XX W CERMAK RD  1012   
12235566            2            RESIDENCE     031XX W 44TH ST   922   
12234792            3            RESIDENCE   107XX S MORGAN ST  2234   
12230182            4                   NA  017XX W BERWYN AVE  2012   

          district  ward  community_area  latitude  longitude  x_coordinate  \
ID                                                                            
12236659        14   1.0            22.0      -1.0       -1.0          -1.0   
12214909        10  24.0            29.0      -1.0       -1.0          -1.0   
12235566         9  15.0            58.0      -1.0       -1.0          -1.0   
12234792        22  34.0            75.0      -1.0       -1.0          -1.0   
12230182        20  40.0            77.0      -1.0       -1.0          -1.0   

          y_coordinate  
ID                      
12236659          -1.0  
12214909          -1.0  
12235566          -1.0  
12234792          -1.0  
12230182          -1.0

In [32]:
##Insert crime table
crime_df_sql.to_sql('crime', # The table to load
          engine,             # The engine created above
          schema= username,   # The schema where the table lives, our pawprint
          if_exists='append', # If the table is found, it would keep loading the end of table.
          index=False,        # Recall that panda data frame has a row index, so we need to ignore it
          chunksize=100)       # Do 100 records from the data frame at a time

2020-12-18 10:06:06,983 INFO sqlalchemy.engine.base.Engine select relname from pg_class c join pg_namespace n on n.oid=c.relnamespace where n.nspname=%(schema)s and relname=%(name)s
2020-12-18 10:06:06,984 INFO sqlalchemy.engine.base.Engine {'schema': 'dlr6w3', 'name': 'crime'}
2020-12-18 10:06:06,989 INFO sqlalchemy.engine.base.Engine BEGIN (implicit)
2020-12-18 10:06:06,991 INFO sqlalchemy.engine.base.Engine INSERT INTO dlr6w3.crime (crime_id, iucr, fbi, primary_desc, description) VALUES (%(crime_id)s, %(iucr)s, %(fbi)s, %(primary_desc)s, %(description)s)
2020-12-18 10:06:06,991 INFO sqlalchemy.engine.base.Engine ({'crime_id': 0, 'iucr': '0266', 'fbi': '02', 'primary_desc': 'CRIMINAL SEXUAL ASSAULT', 'description': 'PREDATORY'}, {'crime_id': 1, 'iucr': '0620', 'fbi': '05', 'primary_desc': 'BURGLARY', 'description': 'UNLAWFUL ENTRY'}, {'crime_id': 2, 'iucr': '1752', 'fbi': '17', 'primary_desc': 'OFFENSE INVOLVING CHILDREN', 'description': 'AGGRAVATED CRIMINAL SEXUAL ABUSE BY FAMILY ME

In [33]:
##Insert location table
location_df_sql.to_sql('location', # The table to load
          engine,             # The engine created above
          schema= username,   # The schema where the table lives, our pawprint
          if_exists='append', # If the table is found, it would keep loading the end of table.
          index=False,        # Recall that panda data frame has a row index, so we need to ignore it
          chunksize=10000)       # Do 10000 records from the data frame at a time

2020-12-18 10:06:13,891 INFO sqlalchemy.engine.base.Engine select relname from pg_class c join pg_namespace n on n.oid=c.relnamespace where n.nspname=%(schema)s and relname=%(name)s
2020-12-18 10:06:13,892 INFO sqlalchemy.engine.base.Engine {'schema': 'dlr6w3', 'name': 'location'}
2020-12-18 10:06:14,000 INFO sqlalchemy.engine.base.Engine BEGIN (implicit)
2020-12-18 10:06:14,097 INFO sqlalchemy.engine.base.Engine INSERT INTO dlr6w3.location (location_id, location_description, block, beat, district, ward, community_area, latitude, longitude, x_coordinate, y_coordinate) VALUES (%(location_id)s, %(location_description)s, %(block)s, %(beat)s, %(district)s, %(ward)s, %(community_area)s, %(latitude)s, %(longitude)s, %(x_coordinate)s, %(y_coordinate)s)
2020-12-18 10:06:14,098 INFO sqlalchemy.engine.base.Engine ({'location_id': 0, 'location_description': 'RESIDENCE', 'block': '028XX W BELDEN AVE', 'beat': 1414, 'district': 14, 'ward': 1.0, 'community_area': 22.0, 'latitude': -1.0, 'longitude':

2020-12-18 10:06:21,124 INFO sqlalchemy.engine.base.Engine INSERT INTO dlr6w3.location (location_id, location_description, block, beat, district, ward, community_area, latitude, longitude, x_coordinate, y_coordinate) VALUES (%(location_id)s, %(location_description)s, %(block)s, %(beat)s, %(district)s, %(ward)s, %(community_area)s, %(latitude)s, %(longitude)s, %(x_coordinate)s, %(y_coordinate)s)
2020-12-18 10:06:21,124 INFO sqlalchemy.engine.base.Engine ({'location_id': 30000, 'location_description': 'RESIDENCE', 'block': '078XX S ELLIS AVE', 'beat': 624, 'district': 6, 'ward': 8.0, 'community_area': 69.0, 'latitude': 41.752390498000004, 'longitude': -87.60036227799999, 'x_coordinate': 1184252.0, 'y_coordinate': 1853214.0}, {'location_id': 30001, 'location_description': 'STREET', 'block': '013XX W 31ST ST', 'beat': 923, 'district': 9, 'ward': 11.0, 'community_area': 60.0, 'latitude': 41.837873151, 'longitude': -87.65875559, 'x_coordinate': 1168074.0, 'y_coordinate': 1884233.0}, {'locati

2020-12-18 10:06:27,926 INFO sqlalchemy.engine.base.Engine INSERT INTO dlr6w3.location (location_id, location_description, block, beat, district, ward, community_area, latitude, longitude, x_coordinate, y_coordinate) VALUES (%(location_id)s, %(location_description)s, %(block)s, %(beat)s, %(district)s, %(ward)s, %(community_area)s, %(latitude)s, %(longitude)s, %(x_coordinate)s, %(y_coordinate)s)
2020-12-18 10:06:27,926 INFO sqlalchemy.engine.base.Engine ({'location_id': 60000, 'location_description': 'SIDEWALK', 'block': '034XX N HALSTED ST', 'beat': 1924, 'district': 19, 'ward': 44.0, 'community_area': 6.0, 'latitude': 41.943844128, 'longitude': -87.649380278, 'x_coordinate': 1170321.0, 'y_coordinate': 1922869.0}, {'location_id': 60001, 'location_description': 'SIDEWALK', 'block': '018XX W NORTH AVE', 'beat': 1434, 'district': 14, 'ward': 32.0, 'community_area': 24.0, 'latitude': 41.910534535, 'longitude': -87.674211066, 'x_coordinate': 1163659.0, 'y_coordinate': 1910678.0}, {'location

2020-12-18 10:06:34,741 INFO sqlalchemy.engine.base.Engine INSERT INTO dlr6w3.location (location_id, location_description, block, beat, district, ward, community_area, latitude, longitude, x_coordinate, y_coordinate) VALUES (%(location_id)s, %(location_description)s, %(block)s, %(beat)s, %(district)s, %(ward)s, %(community_area)s, %(latitude)s, %(longitude)s, %(x_coordinate)s, %(y_coordinate)s)
2020-12-18 10:06:34,742 INFO sqlalchemy.engine.base.Engine ({'location_id': 90000, 'location_description': 'RESIDENCE', 'block': '022XX N LA CROSSE AVE', 'beat': 2522, 'district': 25, 'ward': 31.0, 'community_area': 19.0, 'latitude': 41.92145852, 'longitude': -87.74746301200001, 'x_coordinate': 1143693.0, 'y_coordinate': 1914514.0}, {'location_id': 90001, 'location_description': 'STREET', 'block': '031XX W 61ST ST', 'beat': 823, 'district': 8, 'ward': 15.0, 'community_area': 66.0, 'latitude': 41.782618389, 'longitude': -87.702449716, 'x_coordinate': 1156315.0, 'y_coordinate': 1864008.0}, {'locat

2020-12-18 10:06:41,511 INFO sqlalchemy.engine.base.Engine INSERT INTO dlr6w3.location (location_id, location_description, block, beat, district, ward, community_area, latitude, longitude, x_coordinate, y_coordinate) VALUES (%(location_id)s, %(location_description)s, %(block)s, %(beat)s, %(district)s, %(ward)s, %(community_area)s, %(latitude)s, %(longitude)s, %(x_coordinate)s, %(y_coordinate)s)
2020-12-18 10:06:41,512 INFO sqlalchemy.engine.base.Engine ({'location_id': 120000, 'location_description': 'APARTMENT', 'block': '027XX W CERMAK RD', 'beat': 1034, 'district': 10, 'ward': 28.0, 'community_area': 30.0, 'latitude': 41.851912261, 'longitude': -87.694663208, 'x_coordinate': 1158251.0, 'y_coordinate': 1889274.0}, {'location_id': 120001, 'location_description': 'RESIDENCE', 'block': '113XX S WALLACE ST', 'beat': 2233, 'district': 22, 'ward': 34.0, 'community_area': 49.0, 'latitude': 41.688027804, 'longitude': -87.637404534, 'x_coordinate': 1174334.0, 'y_coordinate': 1829677.0}, {'loc

2020-12-18 10:06:48,101 INFO sqlalchemy.engine.base.Engine INSERT INTO dlr6w3.location (location_id, location_description, block, beat, district, ward, community_area, latitude, longitude, x_coordinate, y_coordinate) VALUES (%(location_id)s, %(location_description)s, %(block)s, %(beat)s, %(district)s, %(ward)s, %(community_area)s, %(latitude)s, %(longitude)s, %(x_coordinate)s, %(y_coordinate)s)
2020-12-18 10:06:48,102 INFO sqlalchemy.engine.base.Engine ({'location_id': 150000, 'location_description': 'DEPARTMENT STORE', 'block': '017XX W HOWARD ST', 'beat': 2422, 'district': 24, 'ward': 49.0, 'community_area': 1.0, 'latitude': 42.019399236999995, 'longitude': -87.675049485, 'x_coordinate': 1163126.0, 'y_coordinate': 1950347.0}, {'location_id': 150001, 'location_description': 'SIDEWALK', 'block': '0000X N LECLAIRE AVE', 'beat': 1533, 'district': 15, 'ward': 28.0, 'community_area': 25.0, 'latitude': 41.880760246, 'longitude': -87.752630735, 'x_coordinate': 1142387.0, 'y_coordinate': 1899

2020-12-18 10:06:55,034 INFO sqlalchemy.engine.base.Engine INSERT INTO dlr6w3.location (location_id, location_description, block, beat, district, ward, community_area, latitude, longitude, x_coordinate, y_coordinate) VALUES (%(location_id)s, %(location_description)s, %(block)s, %(beat)s, %(district)s, %(ward)s, %(community_area)s, %(latitude)s, %(longitude)s, %(x_coordinate)s, %(y_coordinate)s)
2020-12-18 10:06:55,035 INFO sqlalchemy.engine.base.Engine ({'location_id': 180000, 'location_description': 'SIDEWALK', 'block': '042XX W MONROE ST', 'beat': 1115, 'district': 11, 'ward': 28.0, 'community_area': 26.0, 'latitude': 41.879690716999995, 'longitude': -87.73145255799999, 'x_coordinate': 1148157.0, 'y_coordinate': 1899324.0}, {'location_id': 180001, 'location_description': 'VEHICLE NON-COMMERCIAL', 'block': '015XX S DRAKE AVE', 'beat': 1021, 'district': 10, 'ward': 24.0, 'community_area': 29.0, 'latitude': 41.860303214, 'longitude': -87.714024796, 'x_coordinate': 1152954.0, 'y_coordina

2020-12-18 10:07:01,744 INFO sqlalchemy.engine.base.Engine INSERT INTO dlr6w3.location (location_id, location_description, block, beat, district, ward, community_area, latitude, longitude, x_coordinate, y_coordinate) VALUES (%(location_id)s, %(location_description)s, %(block)s, %(beat)s, %(district)s, %(ward)s, %(community_area)s, %(latitude)s, %(longitude)s, %(x_coordinate)s, %(y_coordinate)s)
2020-12-18 10:07:01,745 INFO sqlalchemy.engine.base.Engine ({'location_id': 210000, 'location_description': 'STREET', 'block': '007XX S KEDZIE AVE', 'beat': 1134, 'district': 11, 'ward': 24.0, 'community_area': 27.0, 'latitude': 41.872553245, 'longitude': -87.70587663200001, 'x_coordinate': 1155141.0, 'y_coordinate': 1896773.0}, {'location_id': 210001, 'location_description': 'STREET', 'block': '002XX E 68TH ST', 'beat': 322, 'district': 3, 'ward': 20.0, 'community_area': 69.0, 'latitude': 41.771079015, 'longitude': -87.6195163, 'x_coordinate': 1178969.0, 'y_coordinate': 1859980.0}, {'location_i

2020-12-18 10:07:08,794 INFO sqlalchemy.engine.base.Engine INSERT INTO dlr6w3.location (location_id, location_description, block, beat, district, ward, community_area, latitude, longitude, x_coordinate, y_coordinate) VALUES (%(location_id)s, %(location_description)s, %(block)s, %(beat)s, %(district)s, %(ward)s, %(community_area)s, %(latitude)s, %(longitude)s, %(x_coordinate)s, %(y_coordinate)s)
2020-12-18 10:07:08,795 INFO sqlalchemy.engine.base.Engine ({'location_id': 240000, 'location_description': 'RESIDENCE', 'block': '065XX S MAY ST', 'beat': 724, 'district': 7, 'ward': 17.0, 'community_area': 68.0, 'latitude': 41.775347515, 'longitude': -87.65317514600001, 'x_coordinate': 1169775.0, 'y_coordinate': 1861461.0}, {'location_id': 240001, 'location_description': 'STREET', 'block': '018XX N LOCKWOOD AVE', 'beat': 2532, 'district': 25, 'ward': 37.0, 'community_area': 25.0, 'latitude': 41.91387596, 'longitude': -87.75827156, 'x_coordinate': 1140770.0, 'y_coordinate': 1911731.0}, {'locati

In [34]:
creport_df_sql.to_sql('crime_report', # The table to load
          engine,             # The engine created above
          schema= username,   # The schema where the table lives, our pawprint
          if_exists='append', # If the table is found, it would keep loading the end of table.
          index=False,        # Recall that panda data frame has a row index, so we need to ignore it
          chunksize=100000)       # Do 100000 records from the data frame at a time

2020-12-18 10:08:26,023 INFO sqlalchemy.engine.base.Engine select relname from pg_class c join pg_namespace n on n.oid=c.relnamespace where n.nspname=%(schema)s and relname=%(name)s
2020-12-18 10:08:26,024 INFO sqlalchemy.engine.base.Engine {'schema': 'dlr6w3', 'name': 'crime_report'}
2020-12-18 10:08:26,148 INFO sqlalchemy.engine.base.Engine BEGIN (implicit)
2020-12-18 10:08:27,070 INFO sqlalchemy.engine.base.Engine INSERT INTO dlr6w3.crime_report (case_id, case_number, crime_id, location_id, arrest, domestic, case_date, case_year, updated_on) VALUES (%(case_id)s, %(case_number)s, %(crime_id)s, %(location_id)s, %(arrest)s, %(domestic)s, %(case_date)s, %(case_year)s, %(updated_on)s)
2020-12-18 10:08:27,071 INFO sqlalchemy.engine.base.Engine ({'case_id': 0, 'case_number': 'JD449284', 'crime_id': 0, 'location_id': 0, 'arrest': False, 'domestic': False, 'case_date': datetime.datetime(2012, 11, 27, 18, 0), 'case_year': 2012, 'updated_on': datetime.datetime(2020, 12, 5, 15, 46, 44)}, {'case

2020-12-18 10:09:45,635 INFO sqlalchemy.engine.base.Engine INSERT INTO dlr6w3.crime_report (case_id, case_number, crime_id, location_id, arrest, domestic, case_date, case_year, updated_on) VALUES (%(case_id)s, %(case_number)s, %(crime_id)s, %(location_id)s, %(arrest)s, %(domestic)s, %(case_date)s, %(case_year)s, %(updated_on)s)
2020-12-18 10:09:45,636 INFO sqlalchemy.engine.base.Engine ({'case_id': 300000, 'case_number': 'HV573276', 'crime_id': 16, 'location_id': 222198, 'arrest': False, 'domestic': False, 'case_date': datetime.datetime(2012, 11, 22, 16, 0), 'case_year': 2012, 'updated_on': datetime.datetime(2018, 2, 10, 15, 50, 1)}, {'case_id': 300001, 'case_number': 'HV573265', 'crime_id': 87, 'location_id': 222199, 'arrest': False, 'domestic': False, 'case_date': datetime.datetime(2012, 11, 23, 0, 56), 'case_year': 2012, 'updated_on': datetime.datetime(2018, 2, 10, 15, 50, 1)}, {'case_id': 300002, 'case_number': 'HV573169', 'crime_id': 16, 'location_id': 222200, 'arrest': False, 'do

## For each of the tables, show the number of rows in the table using a sql query


## Crime Table
dsa_student=> select COUNT(*) from crime;
 count
-------
   339
(1 row)

## Location Table
dsa_student=> select COUNT(*) from location;
 count
--------
 244517
(1 row)

## Crime Report Table
dsa_student=> select COUNT(*) from crime_report;
 count
--------
 336149
(1 row)

## <center> Part-II ends here</center>